valhalla/midgard/openlr.h

```diff
@@ -209,41 +258,48 @@ struct LineLocation {
     result.push_back(((first.lfrcnp & 0x7) << 5) | (bearing2integer(first.bearing) & 0x1f));
     result.push_back(distance2integer(first.distance));
 
-    for (const auto& lrp : intermediate) {
+    // Subsequent location reference points are offset lon lat from the first
+    for (auto lrp = std::next(lrps.begin()); lrp != std::prev(lrps.end()); ++lrp) {
       // First longitude
-      append2(static_cast<std::int32_t>(std::round(100000 * (lrp.longitude - longitude))));
-      append2(static_cast<std::int32_t>(std::round(100000 * (lrp.latitude - latitude))));
-      result.push_back(((lrp.frc & 0x7) << 3) | (lrp.fow & 0x7));
-      result.push_back(((lrp.lfrcnp & 0x7) << 5) | (bearing2integer(lrp.bearing) & 0x1f));
-      result.push_back(distance2integer(lrp.distance));
-
-      longitude = lrp.longitude;
-      latitude = lrp.latitude;
+      append2(static_cast<std::int32_t>(std::round(100000 * (lrp->longitude - longitude))));
+      append2(static_cast<std::int32_t>(std::round(100000 * (lrp->latitude - latitude))));
+      result.push_back(((lrp->frc & 0x7) << 3) | (lrp->fow & 0x7));
+      result.push_back(((lrp->lfrcnp & 0x7) << 5) | (bearing2integer(lrp->bearing) & 0x1f));
+      result.push_back(distance2integer(lrp->distance));
+
+      longitude = lrp->longitude;
+      latitude = lrp->latitude;
     }
 
     // Last location reference point
     const auto pofff = poff != 0.f;
     const auto nofff = noff != 0.f;
+    const auto& last = lrps.back();
     append2(static_cast<std::int32_t>(std::round(100000 * (last.longitude - longitude))));
     append2(static_cast<std::int32_t>(std::round(100000 * (last.latitude - latitude))));
     result.push_back(((last.frc & 0x7) << 3) | (last.fow & 0x7));
     result.push_back((pofff << 6) | (nofff << 5) | (bearing2integer(last.bearing) & 0x1f));
 
     // Offsets
     if (pofff) {
-      result.push_back(static_cast<std::int32_t>(256 * poff / first.distance) & 0xff);
+      result.push_back(static_cast<std::int32_t>(std::round(256 * poff / first.distance)) & 0xff);
     }
 
     if (nofff) {
-      result.push_back(static_cast<std::int32_t>(256 * noff / first.distance) & 0xff);
```

Comment:
same bug again using the the first lrps distance when the negative offset is with respect to the second to last lrp